### ISSUE TITLE
unable to find DesignTimeServicesBuilder in ef core 3.1

### ISSUE THREAD

**Issue by xiaotupansy** [dotnet/efcore#19157]
I used to do ef automatic migrations by using DesignTimeServicesBuilder,IMigrationsCodeGenerator,IOperationReporter in ef core 2.2.
But after upgrading to ef core 3.1, I find these classes are missing under Microsoft.EntityFrameworkCore.Design.Internal and Microsoft.EntityFrameworkCore.Migrations.Design.

But after exploring this repository, I do find these classes.
I'm confused about this.
Is this caused by missing some packages?

**Comment by smitpatel** [comment]
Possibly https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-3.0/breaking-changes#dip

**Comment by xiaotupansy** [comment]
Thanks.
I use IMigrationsCodeGenerator to generate snapshot for dbcontext in my code and do automatic migration.
So any other program way to do this in ef core 3?

**Comment by ajcvickers** [comment]
@xiaotupansy Did you try the mitigation of editing your csproj file, as suggested in the link posted above?

**Comment by xiaotupansy** [comment]
@ajcvickers Yes, here's the code:

    ` <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.0">
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>`

Also try this:
` <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.0">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>`

None of them work.

**Comment by xiaotupansy** [comment]
All I need is to generate snapshot in program without using dotnet cli like dotnet ef command.

**Comment by ajcvickers** [comment]
@xiaotupansy Please post a small project that demonstrates the error you are seeing so that we can investigate.

**Comment by xiaotupansy** [comment]
@ajcvickers I solved this problem.
I have two projects in my solution: Repository and Api.
Repository references ef core.
Api does migration things.
Here's the problem: I didn't reference ef core assembly directly in Api project.
When in netcore 2.2, that works.
But after upgrading to 3.1, VS cannot find DesignTimeServicesBuilder in library.
So I tried one solution by adding reference to Microsoft.EntityFrameworkCore.Design, it fails.
But when I add Microsoft.EntityFrameworkCore directly to Api project, it works!
So I guess it's the complex environment (more than one project) causes this problem.
Here's the code:
`    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.0">
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>`

I suggest that you guys add code for Microsoft.EntityFrameworkCore in document.
